### PR TITLE
Update FluentAssertions package dependencies

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes https://github.com/fluentassertions/fluentassertions/issues/2668

Update the ConfigurationManager dependency to 6.0.1 and System.Threading.Tasks.Extensions to 4.5.4.

## IMPORTANT 

* [x] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [ ] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [ ] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [x] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [x] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

### CODE REVIEW

**Comment by ViktorHofer** [comment]
Looking at the build failures...

File: Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj [hunk 1]
```diff
@@ -31,4 +31,9 @@
   <ItemGroup>
     <ProjectReference Include="..\..\Src\FluentAssertions\FluentAssertions.csproj" />
   </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">
+    <Reference Include="System.Net.Http" />
+  </ItemGroup>
```
Comment (jnyrup) [inline Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj:37]: For my edification, why is this necessary now?

Won't this break consumers of Fluent Assertions not having this set?

Context: we have had quite some trouble with `System.Net.Http`.
See e.g. #2122

Comment (ITaluone) [reply Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj:37]: Could most likely be because of the build failure. The log mentions somewhat related

Comment (dennisdoomen) [reply Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj:37]: What build failure are you referring to?

Comment (IT-VBFK) [reply Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj:37]: https://github.com/fluentassertions/fluentassertions/actions/runs/9546652380/job/26310522701

Comment (ViktorHofer) [reply Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj:37]: Yes, those references are now necessary as the two updated test projects don't automatically reference System.Net.Http.dll anymore. I assume you want the long answer, so here it is :)

The version bump of "System.Threading.Tasks.Extensions" from 4.5.0 to 4.5.4 caused this. With 4.5.4, the System.Threading.Tasks.Extensions.dll assembly doesn't depend on netstandard.dll anymore. The .NET SDK automatically adds netstandard facade assemblies for frameworks which support .NET Standard but which don't have the netstandard.dll assembly inbox (i.e. net461, net462, net47).

Back in 2020 the 4.5.4 package got updated to avoid binding redirects by not depending on netstandard.dll anymore. While this is a positive change (better perf and smaller output directory) and was requested by many customers, it's the reason why some projects now need to manually specify additional references.

Overall, this only affects customers on .NET Framework who target a version between 4.6.1 and 4.7.1 and who depend on types from assemblies that aren't referenced by default, i.e. `HttpResponseMessage` from `System.Net.Http.dll`. Our guidance here has been to add those reference items to the project file (or use VS's add reference UI).

https://github.com/fluentassertions/fluentassertions/pull/2122 lists great reasons why framework dependencies (called inbox) should be preferred and why the PR was submitted. cc @AArnott

Comment (dennisdoomen) [reply Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj:37]: Thanks for the extensive explanation. Do I understand you correctly that this may cause a breaking change for folks that target 4.6.1 - 4.7.1, use `HttpResponseMessage` and were relying on FluentAssertions to bring in that dependency?

Comment (ViktorHofer) [reply Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj:37]: That's my guess, yes. It should only affect projects that don't already reference other libraries that trigger the netstandard compat shims to be included. Might make sense to document that.

Comment (ITaluone) [reply Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj:37]: @jnyrup Or do you mean that the variable `TargetFrameworkIdentifier` isn't set per default?

What about adding it into the `<Choose><When>` construct, like so:

```xml
... 
<When Condition="'$(TargetFramework)' == 'net47'">
  <ItemGroup>
       <Compile Remove="Common/NullConfigurationStore.cs" />
  </ItemGroup>
  <ItemGroup>
       <Reference Include="System.Net.Http" />
  </ItemGroup>
....

```

Comment (lg2de) [reply Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj:37]: > That's my guess, yes. It should only affect projects that don't already reference other libraries that trigger the netstandard compat shims to be included. Might make sense to document that.

It could also be handled by the/a nuspec file to express this dependency when targeting e.g. net461.

Comment (ViktorHofer) [reply Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj:37]: > Or do you mean that the variable TargetFrameworkIdentifier isn't set per default?

The `TargetFrameworkIdentifier` property is set by the .NET SDK and doesn't need to be defined manually. It's available in .targets files or inside the project file for item groups and items. No need to set it yourself.

> It could also be handled by the/a nuspec file to express this dependency when targeting e.g. net461.

System.Net.Http.dll was just an example. There are other assemblies that aren't referenced by default on .NET Framework and need to be manually referenced. I don't think that FluentAssertions should do anything in regards to that.

Comment (dennisdoomen) [reply Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj:37]: > That's my guess, yes. It should only affect projects that don't already reference other libraries that trigger the netstandard compat shims to be included. Might make sense to document that.

In that case this should be added as a warning to the breaking changes section of the `releases.md`

**Review by dennisdoomen — APPROVED** [review]

**Comment by ViktorHofer** [comment]
Hello @dennisdoomen, any update on when we can get this PR merged?

**Comment by ITaluone** [comment]
It is waiting for @jnyrup's review.. As it seems, he is very busy these days :)

File: Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj (hunk 1 above)
Comment (jnyrup) [reply Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj:37]: > > Or do you mean that the variable TargetFrameworkIdentifier isn't set per default?
> 
> The `TargetFrameworkIdentifier` property is set by the .NET SDK and doesn't need to be defined manually. It's available in .targets files or inside the project file for item groups and items. No need to set it yourself.
> 
> > It could also be handled by the/a nuspec file to express this dependency when targeting e.g. net461.
> 
> System.Net.Http.dll was just an example. There are other assemblies that aren't referenced by default on .NET Framework and need to be manually referenced. I don't think that FluentAssertions should do anything in regards to that.

It saddens me a bit if this library can't be used out-of-the-box without potentially fiddling with reference assemblies.
If this isn't needed for net472+ but only for net47 and net471, could we include a reference to System.Net.Http.dll only for those two TFMs and would that be problematic?

Comment (jnyrup) [reply Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj:37]: Great to hear.
And we just need to add reference assemblies explicitly in our test projects as we're using project references?

I notice that the generated nuspec file for the nupkg generated from this PR has
`<frameworkAssembly assemblyName="System.Net.Http" targetFramework=".NETFramework4.7, .NETFramework4.7" />`

where the one in 6.12 has
`<frameworkAssembly assemblyName="System.Net.Http" targetFramework=".NETFramework4.7" />`

Do you know what's going on?

Comment (ViktorHofer) [reply Tests/FluentAssertions.Extensibility.Specs/FluentAssertions.Extensibility.Specs.csproj:37]: > And we just need to add reference assemblies explicitly in our test projects as we're using project references?

Exactly. Consumers of FluentAssertions reference it via the package and automatically receive the necessary compiler references.

> Do you know what's going on?

No, not really. I don't know about this schema or which component exactly generates it. But I don't think it causes any issues, it's probably just more verbose.

**Review by jnyrup — APPROVED** [review]
